File: mandelbrotfp.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

void putrgb(long r, long g, long b) {
  unsigned char buf[3];
  buf[0] = r;
  buf[1] = g;
  buf[2] = b;
  write(1, buf, 3);
}

double escape_iteration_count(double cx, double cy) {
  double zx = 0, zy = 0, zx2 = 0, zy2 = 0;
  unsigned long i = 0;
  unsigned long const iteration_limit = 256;
  while (((zx2 + zy2) < 4.0) && (i < iteration_limit)) {
    double tx, ty;
    zx2 = (zx * zx);
    zy2 = (zy * zy);
    tx = cx + (zx2 - zy2);
    ty = cy + (2 * zx * zy);
    i++;
    zx = tx;
    zy = ty;
  }
  return (i == iteration_limit) ? -1 : i;
}

int main(int argc, char *argv[]) {
  long const width = 1024;
  long const height = 1024;
  long y, x;
  
  printf("P6 %d %d 255\n", width, height);
  fflush(NULL);
  for (y = 0; y < height; y++) {
    for (x = 0; x < width; x++) {
      long i = escape_iteration_count(-2.0 + (x * (4.0 / width)),
				      -2.0 + (y * (4.0 / height)));
      long b = (i == -1) ? 0 : i;
      putrgb(b, b, b);
    }
  }
  return 0;
}

File: mandelbrotfp.ml
(* ocamlopt unix.cmxa mandelbrot.ml *)

let putrgb buf stride x y r g b =
  let i = 3 * (x + (y * stride)) in
  String.set buf i r;
  String.set buf (i+1) g;
  String.set buf (i+2) b;
  ()

let escape_iteration_count cx cy =
  let iteration_limit = 256 in
  let rec loop i zx zy =
    if i >= iteration_limit
    then -1
    else let zx2 = zx *. zx in
	 let zy2 = zy *. zy in
	 if (zx2 +. zy2) >= 4.0
	 then i
	 else let tx = cx +. (zx2 -. zy2) in
	      let ty = cy +. (2.0 *. zx *. zy) in
	      loop (i + 1) tx ty
  in loop 0 0.0 0.0

let main () =
  let width = 1024 in
  let height = 1024 in
  Printf.printf "P6 %d %d 255\n%!" width height;
  let buf = String.create (width * height * 3) in
  let rec yloop y =
    if y >= height
    then ()
    else let rec xloop x =
	   if x >= width
	   then yloop (y + 1)
	   else let i =
		  escape_iteration_count
		    (-2.0 +. ((float_of_int x) *. (4.0 /. (float_of_int width))))
		    (-2.0 +. ((float_of_int y) *. (4.0 /. (float_of_int height))))
		in
		let b = char_of_int (if i == -1 then 0 else i) in
		putrgb buf width x y b b b;
		xloop (x + 1)
	 in xloop 0
  in
  yloop 0;
  let _ = Unix.write Unix.stdout buf 0 (String.length buf) in
  ()

let _ = main ()
